components/scream/src/share/grid/remap/coarsening_remapper.hpp

```diff
@@ -0,0 +1,227 @@
+
+#ifdef KOKKOS_ENABLE_CUDA
+public:
+#endif
+  template<int N>
+  void local_mat_vec (const Field& f_src, const Field& f_tgt) const;
+  void pack_and_send () const;
+  void recv_and_unpack () const;
+
+protected:
+  ekat::Comm            m_comm;
+
+  static constexpr bool MpiOnDev = SCREAM_MPI_ON_DEVICE;
+
+  // If MpiOnDev=true, we can pass device pointers to MPI. Otherwise, we need host mirrors.
+  template<typename T>
+  using mpi_view_1d = typename std::conditional<
+                        MpiOnDev,
+                        view_1d<T>,
+                        typename view_1d<T>::HostMirror
+                      >::type;
+
+  // An "overlapped" tgt grid, that is a version of the tgt grid where
+  // ranks own all rows that are affected by local dofs in their src grid
+  grid_ptr_type         m_ov_tgt_grid;
+
+  // Source, target, and overlapped-target fields
+  std::vector<Field>    m_src_fields;
+  std::vector<Field>    m_ov_tgt_fields;
+  std::vector<Field>    m_tgt_fields;
+
+  // Sparse matrix representation in triplet form
+  view_1d<int>    m_row_offsets;
+  view_1d<int>    m_col_lids;
+  view_1d<Real>   m_weights;
+
+  //  ------- MPI data structures -------- //
+
+  // The send/recv buf for pack/unpack
+  view_1d<Real>         m_send_buffer;
+  view_1d<Real>         m_recv_buffer;
+
+  // The send/recv buf to feed to MPI. If MpiOnDev=true, they alias the ones aboce
+  mpi_view_1d<Real>     m_mpi_send_buffer;
+  mpi_view_1d<Real>     m_mpi_recv_buffer;
+
+  // offset(I,J) = offset in send/recv buffor send/recv to/from pid J for field I
+  view_2d<int>          m_send_f_pid_offsets;
+  view_2d<int>          m_recv_f_pid_offsets;
+
+  // Offset of each PID in the send/recv buffer
+  view_1d<int>          m_send_pid_offsets;
+  view_1d<int>          m_recv_pid_offsets;
+
+  // Reorder the lids so that all lids to send/recv to/from pid N
+  // come before those for pid N+1. Also,
+  //   lids_pids(i,0): ith lid to send to pid=lids_pids(i,1)
+  // Note: send/recv lids are the lids of gids in the ov_tgt/tgt grids.
+  //       But here, dofs are ordered differently, so that all dofs to
+  //       send/recv to/from the same pid are contiguous.
+  view_2d<int>          m_send_lids_pids;
+  view_2d<int>          m_recv_lids_pids;
+
+  // Store the start of lids to send/recv to/from each pid in the views above
+  // Note: these are different form m_[send|recv]_pid_offsets. Those are
+  //       offsets in the full send/recv buffer, while these are offsets in
+  //       the views above.
+  view_1d<int>          m_send_pid_lids_start;
+  view_1d<int>          m_recv_pid_lids_start;
+
+  // Send/recv requests
+  std::vector<MPI_Request>  m_recv_req;
+  std::vector<MPI_Request>  m_send_req;
+
+  // Unfortunately, MPI's startall/waitall needs a ptr to nonconst requests.
+  // Since do_remap_fwd is a const method, calling m_recv_req.data() would
+  // return a pointer to const. So store the nonconst ptrs.
```

Comment:
You could also use the `mutable` keyword on the above vectors here if you preferred.